### PR TITLE
gh-107913: Fix possible losses of OSError error codes

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
Functions like PyErr_SetFromErrno() and SetFromWindowsErr() should be called immediately after using the C API which sets errno or the Windows error code.

* Issue: gh-107913

### CODE REVIEW

File: Modules/_io/fileio.c [hunk 1]
```diff
@@ -393,6 +393,11 @@ _io_FileIO___init___impl(fileio *self, PyObject *nameobj, const char *mode,
 
             if (async_err)
                 goto error;
+
+            if (self->fd < 0) {
+                PyErr_SetFromErrnoWithFilenameObject(PyExc_OSError, nameobj);
+                goto error;
+            }
```
Comment (vstinner) [inline Modules/_io/fileio.c:400]: ``fd_is_own = 1;`` changes the code in the error label, it's unclear to me if your change is correct or not.

File: Modules/_ssl.c [hunk 2]
```diff
@@ -3914,8 +3914,8 @@ _ssl__SSLContext_load_cert_chain_impl(PySSLContext *self, PyObject *certfile,
             /* the password callback has already set the error information */
         }
         else if (errno != 0) {
-            ERR_clear_error();
             PyErr_SetFromErrno(PyExc_OSError);
+            ERR_clear_error();
```
Comment (vstinner) [inline Modules/_ssl.c:3918]: Is this function related to errno? Or is it specific to ssl and it leaves errno unchanged?

File: Modules/mmapmodule.c [hunk 3]
```diff
@@ -1355,16 +1355,19 @@ new_mmap_object(PyTypeObject *type, PyObject *args, PyObject *kwdict)
...
-    }
-
     if (m_obj->data == (char *)-1) {
         m_obj->data = NULL;
-        Py_DECREF(m_obj);
+        /* Should be called immediately after function which set errno. */
         PyErr_SetFromErrno(PyExc_OSError);
+        if (devzero != -1) {
+            close(devzero);
+        }
```
Comment (vstinner) [inline Modules/mmapmodule.c]: Copying these 3 lines sounds overkill. The code would look better if you would just save/restore errno.

File: Modules/posixmodule.c [hunk 4]
```diff
@@ -4139,8 +4141,8 @@ _listdir_windows_no_opendir(path_t *path, PyObject *list)
         int error = GetLastError();
         if (error == ERROR_FILE_NOT_FOUND)
             goto exit;
-        Py_DECREF(list);
-        list = path_error(path);
+        path_error(path);
+        Py_SETREF(list, NULL);
```
Comment (vstinner) [inline Modules/posixmodule.c]: Py_CLEAR() can be used. Same remark below.

Oh wow, ``list = path_error(path);`` code is wild! But it works :-D

File: Modules/posixmodule.c [hunk 5]
```diff
@@ -7312,12 +7317,14 @@ os_spawnv_impl(PyObject *module, int mode, path_t *path, PyObject *argv)
...
     Py_END_ALLOW_THREADS
 
-    free_string_array(argvlist, argc);
 
-    if (spawnval == -1)
-        return posix_error();
-    else
-        return Py_BuildValue(_Py_PARSE_INTPTR, spawnval);
+    if (spawnval == -1) {
+        posix_error();
```
Comment (vstinner) [inline Modules/posixmodule.c:7328]: I would suggest to save/restore errno here.

File: Modules/posixmodule.c [hunk 6]
```diff
@@ -7643,8 +7651,10 @@ os_fork1_impl(PyObject *module)
         /* parent: release the import lock. */
         PyOS_AfterFork_Parent();
     }
-    if (pid == -1)
+    if (pid == -1) {
+        errno = saved_errno;
         return posix_error();
```
Comment (vstinner) [inline Modules/posixmodule.c:7660]: Hum, it sounds just wrong that we cannot just pass the errno as an argument.

File: Modules/posixmodule.c [hunk 7]
```diff
@@ -10607,12 +10626,12 @@ os_readv_impl(PyObject *module, int fd, PyObject *buffers)
...
     } while (n < 0 && errno == EINTR && !(async_err = PyErr_CheckSignals()));
 
-    iov_cleanup(iov, buf, cnt);
     if (n < 0) {
         if (!async_err)
             posix_error();
-        return -1;
+        n = -1;
     }
+    iov_cleanup(iov, buf, cnt);
```
Comment (vstinner) [inline Modules/posixmodule.c]: Would it be possible to leave it where it is and save/restore errno?

File: Modules/posixmodule.c [hunk 8]
```diff
@@ -10928,28 +10950,32 @@ os_sendfile_impl(PyObject *module, int out_fd, int in_fd, PyObject *offobj,
...
             }
-            else {
-                // no data has been sent; upper application is supposed
-                // to retry on EAGAIN or EBUSY
-                return posix_error();
-            }
+            // no data has been sent; upper application is supposed
+            // to retry on EAGAIN or EBUSY
         }
-        return (!async_err) ? posix_error() : NULL;
```
Comment (vstinner) [inline Modules/posixmodule.c:10956]: I would prefer to use return on error. The new code is harder to think about, the control flow is less usual.

File: Modules/posixmodule.c [hunk 9]
```diff
@@ -14585,15 +14614,22 @@ DirEntry_fetch_stat(PyObject *module, DirEntry *self, int follow_symlinks)
...
         Py_END_ALLOW_THREADS
     }
+
+    if (result != 0) {
+        path_object_error(self->path);
+#if defined(MS_WINDOWS)
+        PyMem_Free(path);
+#else
+        Py_DECREF(ub);
+#endif
```
Comment (vstinner) [inline Modules/posixmodule.c]: i dislike the fact that you have to copy/paste the cleanup code. maybe save/restor errno?

File: Modules/posixmodule.c [hunk 10]
```diff
@@ -14784,10 +14820,13 @@ os_DirEntry_inode_impl(DirEntry *self)
...
-        PyMem_Free(path);
 
-        if (result != 0)
-            return path_object_error(self->path);
+        if (result != 0) {
+            path_object_error(self->path);
+            PyMem_Free(path);
+            return NULL;
+        }
+        PyMem_Free(path);
```
Comment (vstinner) [inline Modules/posixmodule.c]: For me, it's easier to reason about PyMem_Free() when it's closer to where it's used, I would prefer to keep it where it is. In the previous code, for me, the scope of "path" is more obvious to me, thanks to the grouped lines.

File: Python/fileutils.c [hunk 11]
```diff
@@ -1812,18 +1811,17 @@ _Py_fopen_obj(PyObject *path, const char *mode)
...
-        return NULL;
     }
+#ifdef MS_WINDOWS
+    PyMem_Free(wpath);
+#else
+    Py_DECREF(bytes);
+#endif
 
-    if (set_inheritable(fileno(f), 0, 1, NULL) < 0) {
+    if (f != NULL && set_inheritable(fileno(f), 0, 1, NULL) < 0) {
```
Comment (vstinner) [inline Python/fileutils.c]: I dislike changes in this file :-( I prefer to exit ASAP on error.

**Review by vstinner — COMMENTED** [review]
Maybe it would be just easier/safer to save the errno and then pass it to the function raising OSError, rather than relying on the magic errno variable.

When it's about exchanging two lines, it's fine. But when you move large cleanup code just to preserve errno, IMO it's not a good tradeoff. I prefer code readability and code easy to follow, to read/audit. If you care about the original errno value, just save it and add an API which takes an error code. It's weird that Python has such API for Windows, but not for Unix :-)

This pattern ``} while (n < 0 && errno == EINTR && !(async_err = PyErr_CheckSignals()));`` is kind of fragile if you can about the *original* errno.

File: Modules/_io/fileio.c (hunk 1 above)
Comment (serhiy-storchaka) [reply Modules/_io/fileio.c:400]: It only has effect if `self->fd >= 0`.

File: Modules/_ssl.c (hunk 2 above)
Comment (serhiy-storchaka) [reply Modules/_ssl.c:3918]: I do not know, but it is better to be on safe side.

File: Modules/mmapmodule.c (hunk 3 above)
Comment (serhiy-storchaka) [reply Modules/mmapmodule.c]: Done.

File: Modules/posixmodule.c (hunk 4 above)
Comment (serhiy-storchaka) [reply Modules/posixmodule.c]: Yes, for some reasons I used code like `Py_SETREF(list, NULL)` in few places when introduced this macro (`Py_CLEAR()` has an additional check and `Py_SETREF()` clearer expressed the intention). But in this case I think we can ignore the difference and simply use `Py_CLEAR()`.

Done.

File: Modules/posixmodule.c (hunk 5 above)
Comment (serhiy-storchaka) [reply Modules/posixmodule.c:7328]: Done.

File: Modules/posixmodule.c (hunk 7 above)
Comment (serhiy-storchaka) [reply Modules/posixmodule.c]: Done.

File: Modules/posixmodule.c (hunk 8 above)
Comment (serhiy-storchaka) [reply Modules/posixmodule.c:10956]: Done.

File: Modules/posixmodule.c (hunk 9 above)
Comment (serhiy-storchaka) [reply Modules/posixmodule.c]: Done.

File: Modules/posixmodule.c (hunk 10 above)
Comment (serhiy-storchaka) [reply Modules/posixmodule.c]: Done.

File: Python/fileutils.c (hunk 11 above)
Comment (serhiy-storchaka) [reply Python/fileutils.c]: Done.

**Review by serhiy-storchaka — COMMENTED** [review]
Thank you for your review comments, Victor. I hope that I addressed them all.

File: Modules/_ssl.c (hunk 2 above)
Comment (vstinner) [reply Modules/_ssl.c:3918]: Ok, I'm fine with this change.

File: Modules/_io/fileio.c (hunk 1 above)
Comment (vstinner) [reply Modules/_io/fileio.c:400]: Ok, I see. This change is non-trivial, I preferred to double check.

File: Modules/posixmodule.c [hunk 12]
```diff
@@ -4243,7 +4245,8 @@ _posix_listdir(path_t *path, PyObject *list)
     }
 
     if (dirp == NULL) {
-        list = path_error(path);
+        path_error(path);
+        list = NULL;
```
Comment (vstinner) [inline Modules/posixmodule.c:4249]: This list **parameter** is like **really** surprising. Would you mind to take the opportunity to convert it to a regular local variable (and maybe always initialize it to NULL at the beginning)?

It seems like in the case, the caller could pass their own list.

I have a personal preference to code like this for the code below:

```
    list = PyList_New(0);
    if (list == NULL) {
        goto exit;
    }
```

**Review by vstinner — APPROVED** [review]
LGTM! Thanks, it's a nice fix.

I'm concerned about the complex async_err/EINTR code path, but this one is complicated and can deserve its own separated PR.
